### PR TITLE
[spec] Specify object lifecycle and related operations

### PR DESCRIPTION
**Pull request by thetorpedodog**
This formalizes the object lifecycle and the various creation, opening, and closing methods we are already using in the tiledbsoma Python implementation.

See #107.

### CODE REVIEW

**Review by bkmartinjr — DISMISSED** [review]
Caveat: I did a very fast review.  But based upon that, it looks good to me.

File: abstract_specification.md [hunk 1]
```diff
@@ -233,68 +233,290 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+| del(string key)                         | Remove the key/value from the collection. Does not delete the underlying object (value). |
+| add_new_collection(string key, ...)     | Creates a new sub-Collection and adds it to this Collection.                             |
+| add_new_dataframe(string key, ...)      | Creates a new DataFrame and adds it to this Collection.                                  |
+| add_new_dense_ndarray(string key, ...)  | Creates a new DenseNDArray and adds it to this Collection.                               |
+| add_new_sparse_ndarray(string key, ...) | Creates a new SparseNDArray and adds it to this Collection.                              |
+
+A SOMA collection also manages the lifecycle of objects directly instantiated by it.
+Objects accessed via getting a Collection element, or objects created with one of the <code>add*new*<var>object_type</var></code> methods are considered "owned" by the collection.
+This allows all the objects to be automatically closed together by [the collection's close operation](#operation-close-collection-types).
```
Comment (mlin) [inline abstract_specification.md:418]:
```suggestion
All such objects will be automatically closed together by [the collection's close operation](#operation-close-collection-types).
```

**Review by mlin — APPROVED** [review]
LGTM, one minor language tightening suggested inline

File: abstract_specification.md (hunk 1 above)
Comment (thetorpedodog) [reply abstract_specification.md]: This is good.

File: abstract_specification.md [hunk 2]
```diff
@@ -233,68 +233,290 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
-| exists(uri) -> bool                     | Return true if object exists and is a SOMADataFrame.                           |
-| get soma_type                           | Returns the constant "SOMADataFrame"                                           |
-| get schema -> Arrow.Schema              | Return data schema, in the form of an Arrow Schema                             |
-| get index_column_names -> [string, ...] | Return index (dimension) column names.                                         |
-| get count -> int                        | Return the number of rows in the SOMADataFrame.                                |
-| read                                    | Read a subset of data from the SOMADataFrame                                   |
-| write                                   | Write a subset of data to the SOMADataFrame                                    |
+| Operation                                | Description                                        |
+| ---------------------------------------- | -------------------------------------------------- |
+| static create(uri, ...) -> SOMADataFrame | Create a SOMADataFrame.                            |
```
Comment (johnkerl) [inline abstract_specification.md:513]: ditto re end w/ period or not

**Review by johnkerl — CHANGES_REQUESTED** [review]

File: abstract_specification.md [hunk 3]
```diff
@@ -233,68 +233,290 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+SOMA objects are open for **exclusively** reading or writing.
+When a SOMA object is open for writing, only metadata, the schema, and information directly derived from the schema may be read.
+Additionally, for Collection objects, the members of the collection are also accessible when the collection is open in write mode.
+For example:
+
+```
+writable_collection = soma_impl.open('file:///some/collection', 'w')
+dataframe = writable_collection['dataframe']  # OK
+dataframe.schema  # OK
+dataframe.read(...)  # This is NOT allowed and will cause an error.
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply abstract_specification.md:251]: For clarity, our internal specs tend to always use the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) standard (i.e., consistently use `NOT`), but given that this doc has not already adopted it, that is probably an aspirational change.

[the remark this answers is not in the document]
Comment (thetorpedodog) [reply abstract_specification.md:251]: an author [OUGHT TO](https://www.rfc-editor.org/rfc/rfc6919#section-4) use the caps consistently

[the remark this answers is not in the document]
Comment (johnkerl) [reply abstract_specification.md:251]: @bkmartinjr TIL about RFC2119! :) 

If we want to go that route, sure.

[the remark this answers is not in the document]
Comment (bkmartinjr) [reply abstract_specification.md:251]: there are very good reasons for it - which is why the IETF uses it (as does many other orgs like W3C).  It is a future decision which should not hold up this PR -- but coincidentally the topic was it was just raised inside CZI (regarding this very doc being unclear).  

[Here is](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md) an example of one of our specs that strongly benefits from it.

[the remark this answers is not in the document]
Comment (johnkerl) [reply abstract_specification.md:251]: I filed #122 -- this is awesome info -- thanks @bkmartinjr !! :)

**Review by thetorpedodog — COMMENTED** [review]
lots of changes coming in hot

File: abstract_specification.md [hunk 4]
```diff
@@ -233,68 +233,290 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+class Collection:
+    @classmethod
+    def open(cls, uri, ...): ...
+```
+
+```
+soma_collection_t soma_collection_create(char *uri, ...) { ... }
+```
+
+Where possible, the lifecycle should be integrated into language-specific resource management frameworks (e.g. context management with `with` in Python, try-with-resources in Java).
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply abstract_specification.md:348]: I continue to disagree but won't block the PR.

**Review by johnkerl — APPROVED** [review]

**Comment by thetorpedodog** [comment]
@bkmartinjr @mlin actually I would appreciate if you all took another look particularly for the description of URI semantics that I added.

@johnkerl I clicked you by accident but if you want to give it another ✅ be my guest

File: abstract_specification.md [hunk 5]
```diff
@@ -122,6 +122,68 @@ SOMACollection is an unordered, `string`-keyed map of values. Values may be any
...
+#### Collection entry URIs
+
+Collections refer to their components by URI.
+A collection may store its reference to an element by **absolute** URI or **relative** URI.
+
+A reference to an **absolute** URI is, as the name implies, absolute.
+When the backing storage of the collection itself is moved, absolute URI references do not change.
+
+A collection can also refer to sub-paths of itself by **relative** URI.
+If the entire directory storing a collection and a child referred to by relative URI is moved, the relative URI references now refer to the changed base URI.
```
Comment (bkmartinjr) [inline abstract_specification.md:134]: you might add some wording here that points out that "relative" is an attribute of the particular storage system, eg. S3 really doesn't have directories, but has path prefixes that serve the same purpose.  TileDB-Cloud shows another bunch of special cases if I understand it right.

So I would point out that relative is optional.

And I would reword this to something like:    Where relative URIs are supported by the storage system, collections referencing children by a relative URI will always resolve the child with respect to the parent (base) URI.  For example, on a Posix file system, if the entire directory storing a collection .... <same as you have>

File: abstract_specification.md [hunk 6]
```diff
@@ -233,68 +233,290 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+### Operation: close (Collection types)
+
+In addition to the details of the [common close operation](#operation-close), there are some semantics specific to collection types.
+
+When a SOMA collection is directly used to instantiate child SOMA objects, those SOMA objects are considered "owned" by the collection, and the collection manages their lifecycle.
+When the collection is closed, it must close any of these child objects before closing itself.
+For example:
+
+```
+root = soma_impl.open('path://netloc/root/collection', 'w')
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply abstract_specification.md:501]: we should eventually move to an IDL-like definition if we want the spec to be durable.  But as John says, that is future work, not immediately needed (IMHO)

**Review by bkmartinjr — APPROVED** [review]
One suggestion for a clarification around relative URI resolution in collections.  Otherwise LGTM

File: abstract_specification.md (hunk 5 above)
Comment (thetorpedodog) [reply abstract_specification.md:137]: Added some description of this, along with language indicating that we treat “fake” hierarchies just like normal hierarchies.

**Review by johnkerl — APPROVED** [review]

**Comment by thetorpedodog** [comment]
@atolopko-czi @aaronwolen since this is a spec change we do want unanimous approvals on this so please take a look. Thanks!

**Comment by johnkerl** [comment]
@thetorpedodog ✅  ✅

**Review by aaronwolen — APPROVED** [review]
This is really great

File: abstract_specification.md [hunk 7]
```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+| del(string key)                                     | Remove the key/value from the collection. Does not delete the underlying object (value). |
+| add_new_collection(string key, ...)                 | Creates a new sub-Collection and adds it to this Collection.                             |
+| add_new_dataframe(string key, ...)                  | Creates a new DataFrame and adds it to this Collection.                                  |
+| add_new_dense_ndarray(string key, ...)              | Creates a new DenseNDArray and adds it to this Collection.                               |
+| add_new_sparse_ndarray(string key, ...)             | Creates a new SparseNDArray and adds it to this Collection.                              |
+
+A SOMA collection also manages the lifecycle of objects directly instantiated by it.
+Objects accessed via getting a Collection element, or objects created with one of the <code>add_new\_<var>object_type</var></code> methods are considered "owned" by the collection.
+All such objects will be automatically closed together by [the collection's close operation](#operation-close-collection-types).
```
Comment (atolopko-czi) [inline abstract_specification.md:483]: Also worth mentioning they inherit the parent's context? Or do we consider that implementation-specific?

File: abstract_specification.md [hunk 8]
```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+# external_object remains open because it is not owned by root.
+```
+
+A user may close a child object before closing the parent object.
+Because `close` is idempotent, this is explicitly allowed.
+Closing the child will _not_ otherwise affect the parent object; it will remain open as usual.
+
+### Operation: set
+
+Adds an entry to the collection, overwriting an existing key if one exists already.
```
Comment (atolopko-czi) [inline abstract_specification.md:542]: Does the spec have an opinion on what happens to data originally stored by the object that is replaced at a specified key? Is it orphaned or removed?

File: abstract_specification.md [hunk 9]
```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+- key - The key to set.
+- value - The value to set the key to. If a user sets a collection entry to a type inconsistent with the type of that key (e.g. a defined key in an Experiment) or with the general type of the collection (e.g. a collection of Measurements), behavior is unspecified. Implementations are encouraged to raise an error if possible, to prevent latent errors when later attempting to use the data.
+- uri_type - How the collection should refer to the URI of the newly-added element, whether by absolute or relative URI. The default is `auto`, which will use a relative URI if possible but otherwise use an absolute URI. If `absolute`, the entry will always use an absolute URI. If `relative`, the entry will always use a relative URI, and an error will be raised if a relative URI cannot be used.
+
+### Operation: add_new\_<var>object_type</var>
+
+Each <code>add_new\_<var>object_type</var></code> method creates a new SOMA dataset in storage, adds it to the collection, and returns it to the user. The newly-created entry has the same `context` value as the existing collection and is [owned by the current collection](#operation-close-collection-types).
```
Comment (atolopko-czi) [inline abstract_specification.md:556]:
```suggestion
Each <code>add_new\_<var>object_type</var></code> method creates a new SOMA dataset in storage, adds it to the collection (with the same semantics as the `create` operation), and returns it to the user. The newly-created entry has the same `context` value as the existing collection and is [owned by the current collection](#operation-close-collection-types).
```

File: abstract_specification.md [hunk 10]
```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+```
+add_new_dataframe(string key, string uri = "", ...) -> SOMADataFrame
+add_new_dense_ndarray(string key, string uri = "", ...) -> SOMADenseNDArray
+add_new_sparse_ndarray(string key, string uri = "", ...) -> SOMASparseNDArray
+```
+
+Parameters:
+
+- key: The key to add the new element at. This cannot already be a key of the collection.
+- uri: An optional parameter to specify a URI to create the new collection, which may be [relative or absolute](#collection-entry-uris). If the URI is relative, the new entry will be added with that relative URI. If the URI is absolute, the new entry will be added with that absolute URI. If a collection already exists at the user-provided URI, the operation should fail. If the user does not specify a URI, the collection will generate a new URI for the entry. When possible, this should be a relative URI based on a sanitized version of the key.
```
Comment (atolopko-czi) [inline abstract_specification.md:567]: Do this need to mention that relative URIs cannot be "multi-level"? I think that's what the spec mentions above.

File: abstract_specification.md [hunk 11]
```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
+- The remaining parameters are passed directly to the respective type's `create` static method, except for `context`, which is always set to the current collection's context.
+
+`add_new_collection` has an extra parameter allowing control over the type of collection added:
+
+```
+add_new_collection(string key, CollectionType type, string uri = "", PlatformConfig platform_config) -> SOMACollection
```
Comment (atolopko-czi) [inline abstract_specification.md:573]: Completely out-of-scope for this PR, but I'm now just wondering why we didn't consider having explicit methods for the various types of collections? `add_new_experiment`, `add_new_measurement`, `add_new_collection`.

File: abstract_specification.md [hunk 12]
```diff
@@ -233,68 +298,298 @@ Any given storage "engine" upon which SOMA is implemented may have additional fe
...
-| read                                    | Read a subset of data from the SOMADataFrame                                   |
-| write                                   | Write a subset of data to the SOMADataFrame                                    |
+| Operation                                | Description                                         |
+| ---------------------------------------- | --------------------------------------------------- |
+| static create(uri, ...) -> SOMADataFrame | Create a SOMADataFrame.                             |
+| get soma_type                            | Returns the constant "SOMADataFrame".               |
+| get schema -> Arrow.Schema               | Return data schema, in the form of an Arrow Schema. |
+| get index_column_names -> [string, ...]  | Return index (dimension) column names.              |
+| get count -> int                         | Return the number of rows in the SOMADataFrame.     |
+| read                                     | Read a subset of data from the SOMADataFrame.       |
```
Comment (atolopko-czi) [inline abstract_specification.md:591]: `Read a subset of data iteratively from the SOMADataFrame`?  Whether or not one is reading a subset, the iterative nature of reads seems like a core part of the design, even at the spec level, given the returns type for DataFrame.read (and SparseNDArray.read())

While most people bothering to read a spec like this will understand that a subset includes the whole set, do we want to rely upon reader knowing that? "Read all or a subset..."

File: abstract_specification.md [hunk 13]
```diff
@@ -465,26 +763,24 @@ Parameters:
...
+| static create(uri, ...) -> SOMASparseNDArray | Create a SOMASparseNDArray named with the URI.                          |
+| get soma_type                                | Returns the constant "SOMASparseNDArray".                               |
+| get shape -> (int, ...)                      | Return length of each dimension, always a list of length `ndim`.        |
+| get ndim -> int                              | Return number of dimensions.                                            |
+| get schema -> Arrow.Schema                   | Return data schema, in the form of an Arrow Schema.                     |
+| get is_sparse -> True                        | Return the constant True.                                               |
+| get nnz -> uint                              | Return the number stored values in the array, including explicit zeros. |
+| read                                         | Read a slice of data from the SOMASparseNDArray.                        |
+| write                                        | Write a slice of data to the SOMASparseNDArray.                         |
```
Comment (atolopko-czi) [inline abstract_specification.md:776]: important to use `slice` here instead of `subset` as we do above? If it's a technical term we explicitly want, ok.

File: abstract_specification.md [hunk 14]
```diff
@@ -608,6 +948,7 @@ Array read operations can return results in a variety of formats. The `SOMABatch
 Summary:
 
 ```
+open(string uri, ...) -> SOMA object      # identify a SOMA object and open it
```
Comment (atolopko-czi) [inline abstract_specification.md:951]:
```suggestion
open(string uri, ...) -> SOMAObject      # identify a SOMA object and open it
```
?

**Review by atolopko-czi — APPROVED** [review]
Really great to read the official details on open/close, read/write, etc.! A number of very minor questions and suggestions. Approving as is, since I found no blockers, but have a look at my comments before merging.

**Review by mlin — APPROVED** [review]

File: abstract_specification.md [hunk 15]
```diff
@@ -122,6 +122,71 @@ SOMACollection is an unordered, `string`-keyed map of values. Values may be any
...
+While the term "relative" is used, only **child** URIs are supported as relative URIs; relative paths like `../sibling-path/sub-entry` are not supported.
+
+Consider a directory tree of a SOMA implementation that uses a file named `.collection-info` to store the URIs of the collection's contents:
+
+- `file:///soma-dataset`
+  - `collection`
+    - `.collection-info`
+      - `absolute`: `file:///soma-dataset/collection/abspath`
+      - `external-absolute`: `file:///soma-dataset/more-data/other-data`
+      - `relative`: `relpath`
```
[the remark this answers is not in the document]
Comment (thetorpedodog) [reply abstract_specification.md]: Updated to represent `.collection-info` as a real file and present its contents differently from other paths.
